File: Lesson 9 Maximum slice problem/MaxSliceSum.py
# KADANE's Algorithm
def solution(A):
    if len(A) == 1:
        return A[0]
    if max(A) < 0:
        return max(A)

    max_ending = 0
    max_slice = float('-inf')
    for x in A:
        max_ending = max(0, max_ending+x)
        max_slice = max(max_slice, max_ending)
    return max_slice
    

def solution(A):
    max_ending = 0
    max_slice = A[0]
    
    for x in A:
        max_ending = max_ending + x
        max_slice = max(max_slice, max_ending)
        max_ending = max(0, max_ending)
    return max_slice
    
    
# Same concept but generalizes for more than one slice
def solution(A):
    max_ending = 0
    max_slice = [A[0]]*len(A)
    
    for i in range(len(A)):
        max_ending = max_ending + A[i]
        max_slice[i] = max_ending
        max_ending = max(0, max_ending)
        
    return max(max_slice)

File: Lesson 9 Maximum slice problem/MaxSliceSum.c
// KADANE's Algorithm
int solution(int A[], int N) {
	if(N == 1)
        return A[0];
	
	// Check if all numbers are negative
    int max = -2147483648;
    for(int i=0;i<N;i++)
    {
        if(max<A[i])
            max = A[i];
    }
    if(max<0)
        return max;
    
	// The code below can't handle array of negative numbers ex: [-10], [-10,-100]
    int max_ending = 0, max_slice = -2147483648;
    for(int i=0;i<N;i++)
    {
        if(max_ending+A[i] > 0)
            max_ending = max_ending+A[i];
        else
            max_ending = 0;
        
        if(max_ending > max_slice)
            max_slice = max_ending;
    }
    return max_slice;
}


int solution(int A[], int N) {
    int max_ending = 0, max_slice = A[0];
	
    for(int i=0;i<N;i++)
    {
        max_ending = max_ending+A[i];

        if(max_ending > max_slice)
            max_slice = max_ending;

        if(max_ending < 0)
            max_ending = 0; 
    }
    return max_slice;
}